### ISSUE TITLE
[Bug Report] Application runtime exception in IotHubTransport.

### ISSUE THREAD

**Issue by semeneganov-tomtom** [Azure/azure-iot-sdk-java#1718]
# Context

- **OS and version used:** Android 11 (API 30)
- **Java runtime used:** 1.8
- **SDK version used:** 2.1.5

## Description of the issue
While working throws runtime exception:
```java
java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.checkForOldMessages(IotHubTransport.java:851)
	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.lambda$sendMessages$0$com-microsoft-azure-sdk-iot-device-transport-IotHubTransport(IotHubTransport.java:734)
	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport$$ExternalSyntheticLambda1.run(Unknown Source:2)
	at java.lang.Thread.run(Thread.java:923)
```

Happens frequently.
MQTT connection is used.

## Code sample exhibiting the issue
Just as sample, nothing wrong, I belive, in here.

```java
public void test() throws IotHubClientException, InterruptedException {
    ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
    DeviceClient device = new DeviceClient("<<ConnectionString>>", IotHubClientProtocol.MQTT);
    device.open(false);
    Runnable task = () -> {
        try {
            device.sendEvent(new Message("poison"));
        } catch (IotHubClientException|InterruptedException e) {
            System.out.println("Send exception: " + e.getMessage());
        }

        try {
            device.open(false);
        } catch (Exception e) {
            System.out.println("Ignore exception");
        }
    };
    executor.scheduleAtFixedRate(task, 0, 10, TimeUnit.MILLISECONDS);

    Thread.sleep(100000);

	// Then turn off/on the connection and wait till it happen
}
```

## Console log of the issue
Remove partly because of legal and privacy.
Obviously they are not going to show you the whole run.

```
05-03 18:30:42.663  4474 17314 I IotSdkDeviceClient: CONNECTION STATUS UPDATE: CONNECTED
05-03 18:30:42.663  4474 17314 I IotSdkDeviceClient: CONNECTION STATUS REASON: CONNECTION_OK
05-03 18:30:42.663  4474 17314 I IotSdkDeviceClient: CONNECTION STATUS ERROR MESSAGE: null
05-03 18:30:42.663  4474 17314 I ConnectionChangeCallback: The connection was successfully established. Can send messages.
05-03 18:30:42.664  4474 17314 I DeviceClient: Device client opened successfully
05-03 18:30:42.664  4474 17314 I DeviceProperties: DeviceProperties: Setup device properties
--------- beginning of crash
05-03 18:30:42.701  4474 17346 E AndroidRuntime: FATAL EXCEPTION: Thread-251
05-03 18:30:42.701  4474 17346 E AndroidRuntime: Process: ______, PID: 4474
05-03 18:30:42.701  4474 17346 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
05-03 18:30:42.701  4474 17346 E AndroidRuntime: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.checkForOldMessages(IotHubTransport.java:851)
05-03 18:30:42.701  4474 17346 E AndroidRuntime: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.lambda$sendMessages$0$com-microsoft-azure-sdk-iot-device-transport-IotHubTransport(IotHubTransport.java:734)
05-03 18:30:42.701  4474 17346 E AndroidRuntime: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport$$ExternalSyntheticLambda1.run(Unknown Source:2)
05-03 18:30:42.701  4474 17346 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:923)
05-03 18:30:42.701  4474 17346 D MonitoringInstr: Handling an uncaught exception thrown on the thread Thread-251.
05-03 18:30:42.701  4474 17346 D MonitoringInstr: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
05-03 18:30:42.701  4474 17346 D MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.checkForOldMessages(IotHubTransport.java:851)
05-03 18:30:42.701  4474 17346 D MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.lambda$sendMessages$0$com-microsoft-azure-sdk-iot-device-transport-IotHubTransport(IotHubTransport.java:734)
05-03 18:30:42.701  4474 17346 D MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport$$ExternalSyntheticLambda1.run(Unknown Source:2)
05-03 18:30:42.701  4474 17346 D MonitoringInstr: 	at java.lang.Thread.run(Thread.java:923)
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: An unhandled exception was thrown by the app.
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.checkForOldMessages(IotHubTransport.java:851)
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.lambda$sendMessages$0$com-microsoft-azure-sdk-iot-device-transport-IotHubTransport(IotHubTransport.java:734)
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport$$ExternalSyntheticLambda1.run(Unknown Source:2)
05-03 18:30:42.702  4474 17346 W AndroidJUnitRunner: 	at java.lang.Thread.run(Thread.java:923)
05-03 18:30:42.709  4474 17346 I AndroidJUnitRunner: Bringing down the entire Instrumentation process.
05-03 18:30:42.710  4474 17346 E MonitoringInstr: Exception encountered by: Thread[Thread-251,5,main]. Dumping thread state to outputs and pining for the fjords.
05-03 18:30:42.710  4474 17346 E MonitoringInstr: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
05-03 18:30:42.710  4474 17346 E MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.checkForOldMessages(IotHubTransport.java:851)
05-03 18:30:42.710  4474 17346 E MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.lambda$sendMessages$0$com-microsoft-azure-sdk-iot-device-transport-IotHubTransport(IotHubTransport.java:734)
05-03 18:30:42.710  4474 17346 E MonitoringInstr: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport$$ExternalSyntheticLambda1.run(Unknown Source:2)
05-03 18:30:42.710  4474 17346 E MonitoringInstr: 	at java.lang.Thread.run(Thread.java:923)
```

## The investigations
Extremely hard to reproduce. Manually I didn't have luck to do it.
But it seems that is because of the concurrency.

What happened in the rows of [IotHubTransport.java:851](https://github.com/Azure/azure-iot-sdk-java/blob/214a54007ede4e23002dc17a3cead02784cd0bf8/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java#L851):
```java
if (System.currentTimeMillis() - correlationStartTimeMillis.get(correlationId) >= DEFAULT_CORRELATION_ID_LIVE_TIME)
```

The only call here can bring null is `correlationStartTimeMillis.get(correlationId)` which can cause the NPE in above code.

And the concurrently running thread that will remove the entry can lead on [this rows](https://github.com/Azure/azure-iot-sdk-java/blob/214a54007ede4e23002dc17a3cead02784cd0bf8/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java#L1213):

```java
new Thread(() ->
	{
		correlationCallbacks.remove(correlationId);
		correlationCallbackContexts.remove(correlationId);
		correlationStartTimeMillis.remove(correlationId);
	}).start();
```

**Comment by timtay-microsoft** [comment]
This does look like a bug. We'll try to get this fixed soon.

**Comment by semeneganov-tomtom** [comment]
Thank you @timtay-microsoft for the quick response to the issue.
But I've got couple of question about it, can you help with them?

1. When approximately I can expect this bug fix to be released?
2. When to close the issue, after release will be done?

**Comment by timtay-microsoft** [comment]
1. I'm hoping to get a release out with this fix within this week
2. I'll close this issue thread once the fix has been released

**Comment by timtay-microsoft** [comment]
The release is now complete. This bug is fixed as of DeviceClient version 2.2.0. See [release notes](https://github.com/Azure/azure-iot-sdk-java/releases/tag/2023-06-06) for additional details.

I'll go ahead and close this thread, but feel free to open a new thread if you encounter any further issues!
